### PR TITLE
Ansible mutating webhook

### PR DESCRIPTION
**Pull request by geoand**
This PR introduces a set of tasks that automate the installation of the mutating webhook needed for automatic sidecar injection (essentially it implements [this](https://github.com/istio/istio/issues/4366) issue).

It was tested by introducing the fix mentioned [here](https://github.com/istio/istio/issues/4368#issuecomment-374505904)

### CODE REVIEW

**Comment by sdake** [comment]
related Helm PR: https://github.com/istio/istio/pull/3785

**Comment by geoand** [comment]
@sdake So the preferred approach to checking whether the webhooks can be installed on not based on the Istio version, but on the presence of a specific API capabilitie? 
If so, I can update the playbook to use it

**Comment by sdake** [comment]
@geoand no, I was linking the various work in this area.  Please see:

https://github.com/istio/istio/pull/3935#issuecomment-374239560

**Comment by sdake** [comment]
@yusuoh There are a couple takes on automating webhooks in the various deployment tooling (This is the Ansible take).  Would you link the implementation of your design for the generalized solution when it is available?

Cheers
-steve

**Comment by geoand** [comment]
@sdake Thank you for the update!

File: install/ansible/istio/tasks/main.yml [hunk 1]
```diff
@@ -21,7 +21,7 @@
 
 - name: Extract server version
   shell: |
-    {{ cmd_path }} version | sed -En "{{'s/kubernetes.*v([[:digit:]]\.[[:digit:]]\.[[:digit:]]).*/\1/p' if cluster_flavour == 'ocp' else 's/Server Version.*GitVersion.*v([[:digit:]]\.[[:digit:]]\.[[:digit:]]).*/\1/p'}}" | head -1
+    {{ cmd_path }} version | sed -En "{{'s/kubernetes.*v([[:digit:]]\.[[:digit:]]\.[[:digit:]]).*/\1/p' if cluster_flavour == 'ocp' else 's/Server Version.*GitVersion.*v([[:digit:]]\.[[:digit:]]\.[[:digit:]]).*/\1/p'}}" | tail -1
```
Comment (sdake) [inline install/ansible/istio/tasks/main.yml]: might submit this as a separate PR while the construction around automating webhooks is sorted out.

Comment (geoand) [reply install/ansible/istio/tasks/main.yml]: @sdake Of course, I do that ASAP

Comment (geoand) [reply install/ansible/istio/tasks/main.yml]: Added: https://github.com/istio/istio/pull/4392

**Comment by cmoulliard** [comment]
Can somebody review this PR please ?

**Comment by christian-posta** [comment]
What's the status of this @geoand ? This ready for another review? Or still adding things?

**Comment by geoand** [comment]
It's ready for review. It was tested extensively today by myself and @cmoulliard, hence today's commits

File: install/ansible/istio/tasks/enable_mutating_webhooks.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
+- name: Download missing webbook certificate scripts
```
Comment (yusuoh) [inline install/ansible/istio/tasks/enable_mutating_webhooks.yml]: After #4466 , we no longer need to generate certs and patch CA bundle manually. It will be done by Istio CA and webhook service itself.

Comment (geoand) [reply install/ansible/istio/tasks/enable_mutating_webhooks.yml]: Cool! What istio version has that feature enabled?

Comment (yusuoh) [reply install/ansible/istio/tasks/enable_mutating_webhooks.yml]: It will be in 0.8

Comment (geoand) [reply install/ansible/istio/tasks/enable_mutating_webhooks.yml]: Thank you!

Other than using the version number, is there any other I can know whether or not such functionality is present in the deployed Istio version?

Comment (yusuoh) [reply install/ansible/istio/tasks/enable_mutating_webhooks.yml]: version number is probably the most reliable way to do that.

**Comment by geoand** [comment]
Cool, thanks!

File: install/ansible/istio/tasks/enable_mutating_webhooks.yml (hunk 2 above)
Comment (geoand) [reply install/ansible/istio/tasks/enable_mutating_webhooks.yml]: Just to be clear, for versions newer than `0.8.0`, the only thing that will need to be done to enable automatic sidecar injection, will be to install the sidecar configmap and the sidecar bundle?

Thank you

Comment (yusuoh) [reply install/ansible/istio/tasks/enable_mutating_webhooks.yml]: Yes.

Comment (geoand) [reply install/ansible/istio/tasks/enable_mutating_webhooks.yml]: Nice, I'll update the PR soon

Comment (geoand) [reply install/ansible/istio/tasks/enable_mutating_webhooks.yml]: PR Updated :)

**Comment by geoand** [comment]
The PR has been updated with latest comments from @yusuoh regarding the upcoming changes in Istio 0.8.0

File: install/ansible/istio/tasks/enable_mutating_webhooks.yml [hunk 3]
```diff
@@ -24,16 +25,30 @@
...
+    - istio_version_to_use < minimum_istio_version_for_webhook_certificates
 
-- name: Create signed certificate
-  shell: "{{ istio_dir }}/install/kubernetes/webhook-create-signed-cert.sh --service istio-sidecar-injector --namespace istio-system --secret sidecar-injector-certs"
 
-- name: Install the sidecar injection configmap
-  shell: "{{ cmd_path }} apply -f {{ istio_dir }}/install/kubernetes/istio-sidecar-injector-configmap-release.yaml"
+- block:
+    - name: Create signed certificate
+      shell: "{{ istio_dir }}/install/kubernetes/webhook-create-signed-cert.sh --service istio-sidecar-injector --namespace istio-system --secret sidecar-injector-certs"
```
Comment (sdake) [inline install/ansible/istio/tasks/enable_mutating_webhooks.yml]: These files will not be present in the release, so this conditional logic can be removed.  If you look in:
https://github.com/istio/istio/tree/master/install/kubernetes

You will see no files are present this conditional logic relies upon.  From my understanding, we don't backport features or improvements so there is no viable way this code would be usable as is.

The goal of 0.8 is to be very stable and serve as an LTS until 1.0 is released.  I'd like the Ansible implementation to be ready to roll for at least deployment of Istio on Kubernetes.

File: install/ansible/istio/tasks/enable_mutating_webhooks.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
+- name: Download missing webbook certificate scripts
+  get_url:
+    url: https://raw.githubusercontent.com/istio/istio/master/install/kubernetes/{{ item }}
```
Comment (sdake) [inline install/ansible/istio/tasks/enable_mutating_webhooks.yml]: these files are not available in master.

**Comment by sdake** [comment]
@geoand I believe the goals of this PR (automating webhook installation) have been solved in a general way in PR: https://github.com/istio/istio/pull/4466.  This is why this PR was on hold.  I am not clear this PR will work with master as is, nor work with the 0.8 release of istio, and I'm not sure what value it adds given the changes to how webhook certificates have been automated.  Once 0.8 is released (5-10 days), 0.8 will be a LTS for Istio until the release of 1.0 (as per @thisisnotapril at the most recent community meeting).

If you agree please close and comment.  If not, please let me know what your goals are with this PR (they are not entirely clear now - I thought it was about automating webhook cert creation which was agreed in general in the community to be solved in a generalized way and communicated here: https://github.com/istio/istio/pull/4389#issuecomment-374627797

Cheers
-steve

**Comment by geoand** [comment]
@sdake The goal of the PR was to allow users to enable Istio's automatic webhook feature for various versions of Istio.

If I understand correctly, for Istio 0.8.0, the feature will be enabled when installing Istio, so absolutely no action will need to taken, including creating the configmap. Is this understanding correct?

For older versions of Istio (which the playbook supports you can specify a version), then the files still need to be patched / used to create the k8s entities.
In order to avoid having to support old versions, we could add a assertion to the playbook to force versions 0.8.0 and up.
What do you think?

**Comment by christian-posta** [comment]
I think that makes sense @geoand 

@sdake ?

**Comment by geoand** [comment]
@sdake Supporting older versions should be fairly easy in the code, but I suspect the question is more related to whether we should allow users to install old non-LTS versions in such a way...

I'm open to going either way, so I will wait on more input for Istio members

**Comment by sdake** [comment]
/assign @costinm 

Costin,

Not sure what the policy here should be.  The question is "Do we permit old versions of Istio to be installed with new versions of deployment tools in the Istio repository?"  I could see how old version maintenance could get out of hand.
